### PR TITLE
Fix tests and re-enable test suite

### PR DESCRIPTION
**Pull request by ChristofKaufmann**
# Description

This fixes the tests and re-enables the test suite. The test suite was failing due to a segmentation fault before, see #16 and especially the discussion in #27. This could be resolved by adding the yum package `mesa-libGL`. After that some tests failed, due to different reasons (missing dependencies, wrong versions of dependencies, bugs). The MuJoCo tests required another yum package: `mesa-libOSMesa` and setting the environment variable `MUJOCO_GL=osmesa` to enable offscreen rendering.

The package builds locally for linux64, python 3.9 without errors now. The test stats are:

- 1706 passed
- 2 skipped
- 74 warnings

## Notes

Since the test suite is not applied to the latest release of `gymnasium`, but to 0.27.1, some changes in the recipe are anticipated and marked by comments in the recipe. This should make updates to 0.28.* easier. These are:

- `0001-add-missing-phys2d-asset-in-setup.patch`
  Is required for the current release 0.28.1, but an [upstream pull request](https://github.com/Farama-Foundation/Gymnasium/pull/454) is ~about to be~ merged.
- `0002-fix-compatibility-version-mismatch.patch`
  Is only required for <0.28.0 and has been fixed in https://github.com/Farama-Foundation/Gymnasium/pull/335. We still like to use `shimmy` 0.2.1 to avoid other issues.
- `jax-jumpy <1.0.0`
  Is required for <0.28.0 and has been fixed in https://github.com/Farama-Foundation/Gymnasium/pull/345.

## Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by pseudo-rnd-thoughts** [comment]
Amazing job on this

Looking through the issues I found the following issues, 
```
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=692804&view=logs&jobId=9a864fd9-6c8f-52ca-79ce-2aa6dca1a1de&j=3a25022a-f08c-5c04-be7d-d279a7fc48ef&t=03fcd6e6-009d-576b-db27-27f5032302f9

The reported errors are:
- Encountered problems while solving:
-   - nothing provides requested pytorch-cpu >=1.0
-   - nothing provides jaxlib >=0.3.22 needed by jax-0.3.24-pyhd8ed1ab_0
```
Im quite confused by this as both [jax](https://anaconda.org/conda-forge/jax/files) and [pytorch](https://anaconda.org/conda-forge/pytorch-cpu/files) have conda file that match this requirement

For mac OSX tests, there are a couple of issues 
```
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=692804&view=logs&jobId=9a864fd9-6c8f-52ca-79ce-2aa6dca1a1de&j=9c5ef928-2cd6-52e5-dbe6-9d173a7d951b&t=20c71c51-4b27-578b-485d-06ade2de1d00

=========================== short test summary info ============================
ERROR tests/envs/test_action_dim_check.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
ERROR tests/envs/test_envs.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
ERROR tests/envs/test_make.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
ERROR tests/envs/test_mujoco.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
ERROR tests/envs/test_rendering.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
ERROR tests/experimental/wrappers/test_clip_reward.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
ERROR tests/vector/test_sync_vector_env.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
ERROR tests/wrappers/test_autoreset.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
ERROR tests/wrappers/test_order_enforcing.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
ERROR tests/wrappers/test_passive_env_checker.py - RuntimeError: invalid value for environment variable MUJOCO_GL: osmesa
!!!!!!!!!!!!!!!!!!! Interrupted: 10 errors during collection !!!!!!!!!!!!!!!!!!!
======================= 6 warnings, 10 errors in 56.68s ========================
```
This github discussion might help with the issue - https://github.com/deepmind/dm_control/issues/207

**Comment by h-vetinari** [comment]
Sorry, just catching up to the notifications from this. Let me know if you need something, but great progress, thanks a lot!

**Comment by ChristofKaufmann** [comment]
The latest commit fixes a few issues, but no test. The remaining issues form two groups:

## Python 3.11

The issue persists, that the [test environment for `gymnasium-all`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=695509&view=logs&j=9a864fd9-6c8f-52ca-79ce-2aa6dca1a1de&t=10fc5aa2-324e-5982-4c88-6b31fcab16b3&l=1953) cannot be installed:

> package gymnasium-box2d-0.27.1-py311h38be061_1 requires gymnasium 0.27.1 py311h781c19f_1, but none of the providers can be installed

However, in the [test environment for `gymnasium-box2d`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=695509&view=logs&j=9a864fd9-6c8f-52ca-79ce-2aa6dca1a1de&t=10fc5aa2-324e-5982-4c88-6b31fcab16b3&l=1437) it worked:

>     gymnasium:                 0.27.1-py311h781c19f_1     local      
>     gymnasium-box2d:           0.27.1-py311h38be061_1     local      

The outputs are taken from `linux_64_numpy1.23python3.11.____cpython`, but it is the same for all except

- for `osx_64_numpy1.23python3.11.____cpython`, where the [error message](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=695509&view=logs&j=15f8dc28-9f6f-5e3a-5b3c-2b169071e5be&t=63fa899e-c160-5185-75c4-c5fd30c90374&l=1817) is "package gymnasium-toy_text-0.27.1-py311h6eed73b_1 requires python >=3.11,<3.12.0a0, but none of the providers can be installed", although it worked in the test of `gymnasium-toy_text` before and
- for `osx_arm64_numpy1.23python3.11.____cpython`, since there are no tests.

I could need some help here. This only happens for Python 3.11 (or NumPy 1.23).

## MuJoCo OpenGL Errors

The errors are a bit different, depending on the platform:

- `linux_aarch64` and `linux_ppc64le`: "AttributeError: 'NoneType' object has no attribute 'glGetError'"
- `win_64` and `osx_64`: "mujoco.FatalError: gladLoadGL error"
- `osx_64`: Additionally there is "pygame.error: No hardware accelerated renderers available"

I tried to build the package in a Windows VM and it worked. Sometimes windows with a "CartPole" environment popped up for a second or so. So the tests only do not work in the Azure Pipeline. However, I am not sure how to reproduce this (docker on windows with a windows image?) and if it is possible to use offline rendering.

I assume the situation on macOS would be similar. 

I expected `linux_aarch64` and `linux_ppc64le` to pass, since they should use osmesa as well. Maybe I can fix them. For macOS and windows there might not be a solution for offscreen rendering. Should we skip these tests for macOS and windows?

**Comment by h-vetinari** [comment]
Great progress, we're getting close IMO!

> ### Python 3.11
> 
> The issue persists, that the [test environment for `gymnasium-all`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=695509&view=logs&j=9a864fd9-6c8f-52ca-79ce-2aa6dca1a1de&t=10fc5aa2-324e-5982-4c88-6b31fcab16b3&l=1953) cannot be installed:

That's because we have no python 3.11 builds of [gym](https://anaconda.org/conda-forge/gym/files):
```
Mamba failed to solve:
 - gymnasium-all 0.27.1.* py311h38be061_1
 - [...]
 - gym
```

> ### MuJoCo OpenGL Errors
> 
> The errors are a bit different, depending on the platform:
>  * `linux_aarch64` and `linux_ppc64le`: "AttributeError: 'NoneType' object has no attribute 'glGetError'"
>  * `win_64` and `osx_64`: "mujoco.FatalError: gladLoadGL error"
>  * `osx_64`: Additionally there is "pygame.error: No hardware accelerated renderers available"

None of these surprise me terribly; OpenGL needs drivers and/or hardware and we necessarily have them in CI at all, in emulation (aarch/ppc) or on non-Linux platforms (win, osx), particularly since anything about mujoco GL depends on linux-isms.

I would just skip those tests that require OpenGL (or ideally, make the upstream testsuite skip them by default if the driver cannot be found or loaded).

> I tried to build the package in a Windows VM and it worked. Sometimes windows with a "CartPole" environment popped up for a second or so. So the tests only do not work in the Azure Pipeline. However, I am not sure how to reproduce this (docker on windows with a windows image?) and if it is possible to use offline rendering.

IIUC, that sounds like things are actually working pretty well, provided the required hardware is there (as you'll have on an actual end-user machine).

**Comment by ChristofKaufmann** [comment]
> That's because we have no python 3.11 builds of [gym](https://anaconda.org/conda-forge/gym/files)

That makes sense. Somehow the message irritated me. I would have expected something like "package gymnasium-all-0.27.1-py311h38be061_1 requires gym, but none of the providers can be installed".

> I would just skip those tests that require OpenGL (or ideally, make the upstream testsuite skip them by default if the driver cannot be found or loaded).

👍
I will also try to get patches 0003 (skip tests using jax or pytorch, if they are missing) and 0004 (skip "fork" multiprocessing on windows) upstream.

**Comment by pseudo-rnd-thoughts** [comment]
Congrats @ChristofKaufmann, I have closed my re-adding testing PR
@h-vetinari I'm happy to merge if you are

@ChristofKaufmann I'm planning on making a release of gymnasium soon v0.28.2 which should include most of the changes that you have included here

**Comment by ChristofKaufmann** [comment]
That's cool to hear! :-) I think for the conda package, we should still build v0.28.0 and v0.28.1, just to make them available.

Maybe a few notes to the last push:

- [Skipping Mujoco tests if OpenGL is not available](https://github.com/conda-forge/gymnasium-feedstock/pull/28/commits/c7944db7203dc56e8858a7fb5960809c1a462e77) should be **safe** here, since we use a conda package of pytorch and even the CPU version. In https://github.com/Farama-Foundation/Gymnasium/pull/480 I had some problems that need to be worked around before that can be merged in.
- [Skipping the pygame tests for macOS](https://github.com/conda-forge/gymnasium-feedstock/pull/28/commits/2a5a7150caf39093e1b2731aea564689e9bd693c) is quite ugly, but maybe okay for now? Maybe you have a better idea.

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:90]: Let's please use a more readable setup for the test skips. I recommend the one used e.g. by arrow or [scipy](https://github.com/conda-forge/scipy-feedstock/blob/main/recipe/meta.yaml). Since `pytest -k` does substring matches, I removed redundant skips (e.g. skipping `test_play_relevant_keys` will automatically also skip `test_play_relevant_keys_with_env_attribute`). Alphabetically ordering the tests helps discover such instances.

```suggestion
        {% set tests_to_skip = "_not_a_real_test" %}
        # add a comment here like "the following tests are skipped [according to selector] because X"
        {% set tests_to_skip = tests_to_skip + " or test_human_rendering" %}                    # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_keyboard_irrelevant_keydown_event" %}  # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_keyboard_keyup_event" %}               # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_keyboard_quit_event" %}                # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_keyboard_relevant_keydown_event" %}    # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_no_error_warnings[env0]" %}            # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_no_error_warnings[env1]" %}            # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_play_loop_real_env" %}                 # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_play_relevant_keys" %}                 # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_pygame_quit_event" %}                  # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_video_size_no_zoom" %}                 # [osx]
        {% set tests_to_skip = tests_to_skip + " or test_video_size_zoom" %}                    # [osx]
        # need to specify opengl driver on linux (this comment is an example; feel free to improve)
        - export MUJOCO_GL="osmesa"  # [linux]
        - pytest -v tests/ -k 'not ({{ tests_to_skip }})'
```

File: recipe/meta.yaml
Comment (h-vetinari) [inline recipe/meta.yaml:85]:
```suggestion
        - jax >= 0.3.24  # [osx or linux64]
        # conda-forge doesn't have pytorch on all platforms/arches yet
        - pytorch >=1.0  # [unix and not ppc64le]
        - gym            # [py<311]
```
For context see https://github.com/conda-forge/pytorch-cpu-feedstock/issues/155. Also, where reasonable, I prefer aligned selectors for better readability.

**Review by h-vetinari — CHANGES_REQUESTED** [review]
This is looking great, thanks a lot!

Just some small issues left, I added suggestions how to resolve them (but you should update at least the sample comments I left)

File: recipe/meta.yaml
Comment (ChristofKaufmann) [reply recipe/meta.yaml:90]: That looks indeed better! Thanks for the suggestion. However, skipping `test_play_relevant_keys` will also skip `test_play_relevant_keys_no_mapping`, which I explicitly added outside the parenthesis to put it back in, because it passes. I will modify your suggestion (somehow) and add it directly to line 104, without another variable, I think.

File: recipe/meta.yaml
Comment (ChristofKaufmann) [reply recipe/meta.yaml:85]: In a local test, mamba could not see, that I needed the CPU variant of PyTorch. 🤔 I will check the job output.

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:91]: In that case you can couple `test_play_relevant_keys` with something else that makes it unique in the serialization of the test name, e.g.
```
        {% set tests_to_skip = tests_to_skip + " or (test_play_relevant_keys and something_else)" %}  # [osx]
```
or if, every aspect (test path, test class, etc.) is the same as with `test_play_relevant_keys_no_mapping`, then the following should work:
```
        # avoid skipping a passing test due to on identical substrings
        {% set tests_to_skip = tests_to_skip + " or (test_play_relevant_keys and not test_play_relevant_keys_no_mapping)" %}  # [osx]
```

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml]: Ah, sorry, I didn't see that this was only for the test section. Generally, the variant being picked up will depend on whether conda can find the CUDA drivers or not. This will not be the case in CI, but might be the case on your machine.

Background: Hypothetically requiring gymnasium to use the CPU variant would be an extreme restriction to impose on all other packages depending on gymnasium, forcing them to forego the benefits of the CUDA variant for all other purposes even if they have the right hardware - we shouldn't do this under basically any circumstances.

If indeed we want to force the CPU variant for testing, you can add an additional line `pytorch =*=cpu*` to enforce the cpu variant, but I don't think this should be necessary.

File: recipe/meta.yaml
Comment (ChristofKaufmann) [reply recipe/meta.yaml:91]: Yes, the latter is true. I was a bit disappointed that in pytest there is no option to specify the full test name.

Your suggestion

```python
{% set tests_to_skip = tests_to_skip + " or (test_play_relevant_keys and not test_play_relevant_keys_no_mapping)" %}  # [osx]
```

looks better than what I was about to be doing:

```python
pytest -v tests/ -k 'not ({{ tests_to_skip }}) or test_play_relevant_keys_no_mapping'
```

because `test_play_relevant_keys_no_mapping` is closer to `test_play_relevant_keys`.

File: recipe/meta.yaml
Comment (ChristofKaufmann) [reply recipe/meta.yaml]: You are right. The CPU variant was installed in the job. Very good.

I don't remember in what situation pytorch selected the CUDA variant. Maybe it was even from PyPI, when I tested something for gymnasium.

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:91]: > Yes, the latter is true. I was a bit disappointed that in pytest there is no option to specify the full test name.

Well, you can specify the full test name (you just have to use it as pytest would serialize it), but if there are other tests that are named exactly the same just by adding some extra characters to the test name, then it would be a huge jump in complexity of pytest's implementation to not capture those as well (and in many situation, this behaviour is desired - so a feature, not a bug).

I'll note though that perfect is definitely the enemy of the good here. I care about enabling the test suite _at all_, not that we'd be skipping one passing test too many on one platform. I'd be more than happy to just use
```
{% set tests_to_skip = tests_to_skip + " or test_play_relevant_keys" %}  # [osx]
```
In fact I prefer it because the extra noise in the recipe is not worth unskipping a test IMO.

If you really want to fix it, the easiest option would be to avoid identical substrings in tests if they do very different thing (ideally upstream; would be possible through a patch, but for that as well, the complexity vs. benefit trade-off is negative).

Comment (h-vetinari) [reply recipe/meta.yaml:91]: > Is there a way in github I can modify your suggested changes?

You could if you become maintainer on this feedstock (then you could edit the suggestion in my comment & commit it) - would you be interested in that (given your apparent interest in gymnasium & great work in this PR)? 🙃 

> I really just want to change your sample comments. Or should I commit your suggestion, amend changes and force push?

Force-pushing is fine, especially if it helps future git archeology. I tend to clean up the commits in my PRs before merging them so that future me will have an easier time understanding why something was done when I have to dig into the git history.

But in this case, you don't even have to do that. You can just copy & paste my suggestion into your checkout locally & commit it. I don't need the attribution that GH generates. That's not to say that there can't be cases where such metadata operations are desirable (and they can of course be replicated also locally using `git commit --author=` and/or `Co-Authored-By: Name <email>` in the commit message).

Comment (ChristofKaufmann) [reply recipe/meta.yaml:91]: > You could if you become maintainer on this feedstock (then you could edit the suggestion in my comment & commit it) - would you be interested in that (given your apparent interest in gymnasium & great work in this PR)?

Thanks a lot for the offer and your kind words, but probably I could not handle the time it takes. My main motivation for this was actually that I need to maintain some installations and installation instructions in a university to install gymnasium-atari and I like installations with mamba much more than with pip. I saw #16 and that shimmy-atari together with the failing tests are the main blocker. Maybe I can also help to upgrade the package version, since I know that some of my patches have to be adapted for that. I might try to bring autorom and pettingzoo also to conda-forge next, if I can find the time for it. So, its not that I am not interested in it, but I cannot make any promises in the long run.

Comment (h-vetinari) [reply recipe/meta.yaml]: > So, its not that I am not interested in it, but I cannot make any promises in the long run.

Maintainership is not a promise to commit time, we're all volunteers here, and no-one can force use to do this unpaid work. But it does give you easier handling and more of a say. Anyway, it's an open invitation. There's never a shortage of things to do, as you discovered. ;-)

**Review by h-vetinari — APPROVED** [review]

**Comment by h-vetinari** [comment]
Thanks a lot, really happy we have test coverage again! 🥳

**Pull request merged**
